arwen/host/execution.go

```diff
@@ -163,59 +160,23 @@ func (host *vmHost) ExecuteOnDestContext(input *vmcommon.ContractCallInput) (*vm
 	storage.PushState()
 	storage.SetAddress(runtime.GetSCAddress())
 
-	gasUsed := uint64(0)
-	defer func() {
-		vmOutput = host.finishExecuteOnDestContext(gasUsed, err)
-	}()
-
 	// Perform a value transfer to the called SC. If the execution fails, this
 	// transfer will not persist.
-	err = output.TransferValueOnly(input.RecipientAddr, input.CallerAddr, input.CallValue)
+	err := output.TransferValueOnly(input.RecipientAddr, input.CallerAddr, input.CallValue)
 	if err != nil {
-		return nil, err
+		return host.finishExecuteOnDestContext(0, err)
```

Comment:
`host.finishExecuteOnDestContext()` will not use the `0` value because `err != nil`. It will call `outputContext.CreateVMOutputInCaseOfError()` directly, which sets `GasRemaining = GasLocked`. `GasLocked` is 0 for direct calls anyway.

This approach currently follows the principle "an error consumes all the gas (except locked gas)".